File: src/main/java/org/albaross/agents4j/extraction/KnowledgeBase.java
package org.albaross.agents4j.extraction;

import org.albaross.agents4j.extraction.data.Rule;

import java.util.Collection;
import java.util.Set;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import static java.util.stream.Collectors.toSet;

public interface KnowledgeBase<A> extends Iterable<Collection<Rule<A>>> {

    int size();

    default boolean isEmpty() {
        return size() == 0;
    }

    Collection<Rule<A>> reason(Set<String> state);

    default Set<A> actions(Set<String> state) {
        return reason(state).stream()
                .map(Rule::getConclusion)
                .collect(toSet());
    }

    void add(Rule<A> rule);

    void remove(Rule<A> rule);

    default void addAll(Collection<Rule<A>> rules) {
        rules.forEach(this::add);
    }

    default void removeAll(Collection<Rule<A>> rules) {
        rules.forEach(this::remove);
    }

    default KnowledgeBase<A> filter(Consumer<KnowledgeBase<A>> filter) {
        filter.accept(this);
        return this;
    }

    default <T> KnowledgeBase<A> filter(BiConsumer<KnowledgeBase<A>, T> filter, T arg) {
        filter.accept(this, arg);
        return this;
    }

    void clear();

    default Stream<Collection<Rule<A>>> stream() {
        return StreamSupport.stream(spliterator(), false);
    }

}

File: src/main/kotlin/collections/KnowledgeBase.kt
package org.albaross.agents4j.extraction

interface KnowledgeBase<A> : Collection<Rule<A>> {

    fun reason(state: State): Collection<Rule<A>>
}

interface HierarchicalKnowledgeBase<A> : KnowledgeBase<A> {

    val levelCount: Int

    fun level(dim: Int): Collection<Rule<A>>

    fun levels(): Iterable<Collection<Rule<A>>>
}

fun <A> HierarchicalKnowledgeBase<A>.format(): String {

    val builder = StringBuilder().append("--------------------\n")

    for (level in this.levels()) {
        for ((state, action, weight) in level) {
            builder.append(state, " ^ ", "T").append(" => ").append(action).append(" [").append(weight, 3).append("]\n")
        }

        builder.append("--------------------\n")
    }

    builder.append(this.size).append(" rules\n")
    return builder.append("--------------------\n").toString()
}